src/transform-sdk/rust/src/record.rs

```diff
@@ -0,0 +1,368 @@
+
+fn read_kv(payload: &[u8]) -> Result<DecodeResult<KeyValuePair>> {
+    let key = varint::read_sized_buffer(payload)?;
+    let payload = &payload[key.read..];
+    let value = varint::read_sized_buffer(payload)?;
+    Ok(DecodeResult {
+        value: (key.value, value.value),
+        read: key.read + value.read,
+    })
+}
+
+/// A zero-copy record header.
+pub struct BorrowedHeader<'a> {
+    key: Option<&'a [u8]>,
+    value: Option<&'a [u8]>,
+}
+
+impl<'a> BorrowedHeader<'a> {
+    /// Create a new header without any copies.
+    pub fn new(key: Option<&'a [u8]>, value: Option<&'a [u8]>) -> Self {
+        Self { key, value }
+    }
+
+    fn read_from_payload(payload: &'a [u8]) -> Result<DecodeResult<Self>> {
+        read_kv(payload).map(|result| result.map(|(key, value)| BorrowedHeader { key, value }))
+    }
+
+    /// Returns the header's key or `None` if there is no key.
+    pub fn key(&self) -> Option<&[u8]> {
+        self.key
+    }
+
+    /// Returns the header's value or `None` if there is no value.
+    pub fn value(&self) -> Option<&[u8]> {
+        self.value
+    }
+
+    /// Clones this header obtaining ownership of the clone.
+    pub fn to_owned(&self) -> RecordHeader {
+        RecordHeader::new(
+            self.key.map(|b| b.to_owned()),
+            self.value.map(|b| b.to_owned()),
+        )
+    }
+}
+
+/// A zero-copy representation of a record within Redpanda.
+///
+/// BorrowedRecords are handed to [`on_record_written`] event handlers as the record that Redpanda
+/// wrote.
+pub struct BorrowedRecord<'a> {
+    key: Option<&'a [u8]>,
+    value: Option<&'a [u8]>,
+    timestamp: SystemTime,
+    headers: Vec<BorrowedHeader<'a>>,
+}
+
+impl<'a> BorrowedRecord<'a> {
+    /// Create a new record without any copies.
+    ///
+    /// NOTE: This method is useful for tests to mock out custom events to your transform function.
+    pub fn new(key: Option<&'a [u8]>, value: Option<&'a [u8]>, timestamp: SystemTime) -> Self {
+        Self {
+            key,
+            value,
+            timestamp,
+            headers: Vec::new(),
+        }
+    }
+
+    /// Create a new record without any copies.
+    ///
+    /// NOTE: This method is useful for tests to mock out custom events to your transform function.
+    pub fn new_with_headers(
+        key: Option<&'a [u8]>,
+        value: Option<&'a [u8]>,
+        timestamp: SystemTime,
+        headers: Vec<BorrowedHeader<'a>>,
+    ) -> Self {
+        Self {
+            key,
+            value,
+            timestamp,
+            headers,
+        }
+    }
+
+    pub(crate) fn read_from_payload(payload: &'a [u8], timestamp: SystemTime) -> Result<Self> {
+        let kv = read_kv(payload)?;
+        let mut headers: Vec<BorrowedHeader<'a>> = Vec::new();
+        let mut payload = &payload[kv.read..];
+        let header_count_result = varint::read(payload)?;
+        payload = &payload[header_count_result.read..];
+        for _ in 0..header_count_result.value {
```

Comment:
move `let mut headers: Vec<BorrowedHeader<'a>> = Vec::new();`
here
as `let mut headers: Vec<BorrowedHeader<'a>> = Vec::with_capacity(header_count_result.value);`

Reply:
or even if you want to avoid having to care about an extra `mut`

```
let headers = (0..header_count_result.value()).map(|_| {
  ...
}).collect();
```

Reply:
nice catch :smile:. I kept the boring loop instead of the map.